packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts

```diff
@@ -0,0 +1,105 @@
+//
+// Copyright 2024 DXOS.org
+// s
+
+import { type MixedEncoding } from 'level-transcoder';
+
+import { type StorageAdapterInterface, type Chunk, type StorageKey } from '@dxos/automerge/automerge-repo';
+import { type MaybePromise } from '@dxos/util';
+
+import { type MySublevel } from './types';
+
+export type LevelDBStorageAdapterParams = {
+  db: MySublevel;
+  callbacks?: StorageCallbacks;
+};
+
+export type StorageCallbacks = {
+  beforeSave?: (path: StorageKey) => MaybePromise<void>;
+  afterSave?: (path: StorageKey) => MaybePromise<void>;
+};
+
+export class LevelDBStorageAdapter implements StorageAdapterInterface {
+  private _state: 'opened' | 'closed' = 'opened';
+  constructor(private readonly _params: LevelDBStorageAdapterParams) {}
+
+  async load(keyArray: StorageKey): Promise<Uint8Array | undefined> {
+    if (this._state !== 'opened') {
+      return undefined;
+    }
+    return this._params.db
+      .get<StorageKey, Uint8Array>(keyArray, { ...encodingOptions })
+      .catch((err) => (err.code === 'LEVEL_NOT_FOUND' ? undefined : Promise.reject(err)));
+  }
+
+  async save(keyArray: StorageKey, binary: Uint8Array): Promise<void> {
+    if (this._state !== 'opened') {
+      return undefined;
+    }
+    await this._params.callbacks?.beforeSave?.(keyArray);
+    await this._params.db.put<StorageKey, Uint8Array>(keyArray, Buffer.from(binary), {
+      ...encodingOptions,
+    });
+    await this._params.callbacks?.afterSave?.(keyArray);
+  }
+
+  async remove(keyArray: StorageKey): Promise<void> {
+    if (this._state !== 'opened') {
+      return undefined;
+    }
+    await this._params.db.del<StorageKey>(keyArray, { ...encodingOptions });
+  }
+
+  async loadRange(keyPrefix: StorageKey): Promise<Chunk[]> {
+    if (this._state !== 'opened') {
+      return [];
+    }
+    const result: Chunk[] = [];
+    for await (const [key, value] of this._params.db.iterator<StorageKey, Uint8Array>({
```

Comment:
there's an `db.iterator(...).all()` method which is more efficient if you do not intend to process results 1-by-1